src/com/phonebookservice/util/MyArrayList.java

```diff
@@ -340,7 +366,17 @@ public boolean retainAll(final Collection<?> collection) {
      */
     @Override
     public boolean containsAll(final Collection<?> collection) {
-        throw new UnsupportedOperationException();
+        this.checkCollectionNotNull(collection);
+        boolean changed = true;
+        final Iterator<?> iterator = collection.iterator();
+
+        while (iterator.hasNext()) {
+            if (!this.contains(iterator.next())) {
```

Comment:
If you know it is false and won't be changed again to true, why do you continue the loop?

Why don't return false instead of changed = false

Reply:
you are right

Reply:
Please resolve this

Reply:
contains all complexity is n*m 
we can make it better by: 
1) convert list of data into set.
2) save size of set
3) populate set with collection
4) compare final size with one saved, if not equal, then return false
this case complexity is: n + m 
or we can use Map (an idea)  (what I understand that set use map internally)